{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8b891b75",
   "metadata": {},
   "source": [
    "# GFF Processing Metrics"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c30f63f3",
   "metadata": {},
   "source": [
    "Using a processed GFF file (either from NIH or Ensembl) we can apply some metrics to understand the characteristics of our RNA dataset.  In particular, we are interested in characterizing the number of intermediate states possible for a gene once it has left the DNA.  These intermediate states are relevant to the study of alternative splicing, secondary RNA structure, and transcription time."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9606cab6",
   "metadata": {},
   "source": [
    "# Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "81f7945f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Libraries\n",
    "\n",
    "\n",
    "# Combinations\n",
    "from itertools import combinations\n",
    "\n",
    "# NumPy\n",
    "import numpy as np\n",
    "\n",
    "# Pandas\n",
    "import pandas as pd\n",
    "\n",
    "# Sampling\n",
    "# import random  \n",
    "from random import sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f98caada",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set the data folder.\n",
    "data_folder = '/media/apollo/Samsung_T5/transfer/mayur/annotations/'\n",
    "# data_folder = '/home/mad1188/rvallsamples/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "368db10e",
   "metadata": {},
   "source": [
    "# NIH Processing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4dabb931",
   "metadata": {},
   "source": [
    "First, read in the processed GFF file from NIH."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7842bfce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# NIH file.\n",
    "\n",
    "# Skip the metadata lines (1-9).\n",
    "df = pd.read_csv(\n",
    "    data_folder + 'GCF_000001405.40_GRCh38.p14_genomic.gff.exon.processed',\n",
    "    sep = '\\t'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6bfbf31d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>nih_molecule_accession</th>\n",
       "      <th>source</th>\n",
       "      <th>category</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>information</th>\n",
       "      <th>ID</th>\n",
       "      <th>exon_parent</th>\n",
       "      <th>exon_gbkey</th>\n",
       "      <th>exon_gene</th>\n",
       "      <th>exon_transcript_id</th>\n",
       "      <th>predicted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>17369.0</td>\n",
       "      <td>17436.0</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=exon-NR_106918.1-1;Parent=rna-NR_106918.1;D...</td>\n",
       "      <td>exon-NR_106918.1-1</td>\n",
       "      <td>rna-NR_106918.1</td>\n",
       "      <td>precursor_RNA</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>NR_106918.1</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>17369.0</td>\n",
       "      <td>17391.0</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=exon-MIR6859-1-1;Parent=rna-MIR6859-1;Dbxre...</td>\n",
       "      <td>exon-MIR6859-1-1</td>\n",
       "      <td>rna-MIR6859-1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>17409.0</td>\n",
       "      <td>17431.0</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=exon-MIR6859-1-2-1;Parent=rna-MIR6859-1-2;D...</td>\n",
       "      <td>exon-MIR6859-1-2-1</td>\n",
       "      <td>rna-MIR6859-1-2</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>29774.0</td>\n",
       "      <td>30667.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-XR_007065314.1-1;Parent=rna-XR_0070653...</td>\n",
       "      <td>exon-XR_007065314.1-1</td>\n",
       "      <td>rna-XR_007065314.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>XR_007065314.1</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>30976.0</td>\n",
       "      <td>31093.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-XR_007065314.1-2;Parent=rna-XR_0070653...</td>\n",
       "      <td>exon-XR_007065314.1-2</td>\n",
       "      <td>rna-XR_007065314.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>XR_007065314.1</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1698714</th>\n",
       "      <td>NC_012920.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>14149.0</td>\n",
       "      <td>14673.0</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=exon-ND6-1;Parent=rna-ND6;Dbxref=GeneID:454...</td>\n",
       "      <td>exon-ND6-1</td>\n",
       "      <td>rna-ND6</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>ND6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1698715</th>\n",
       "      <td>NC_012920.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>14674.0</td>\n",
       "      <td>14742.0</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=exon-TRNE-1;Parent=rna-TRNE;Dbxref=GeneID:4...</td>\n",
       "      <td>exon-TRNE-1</td>\n",
       "      <td>rna-TRNE</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>TRNE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1698716</th>\n",
       "      <td>NC_012920.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>14747.0</td>\n",
       "      <td>15887.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-CYTB-1;Parent=rna-CYTB;Dbxref=GeneID:4...</td>\n",
       "      <td>exon-CYTB-1</td>\n",
       "      <td>rna-CYTB</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>CYTB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1698717</th>\n",
       "      <td>NC_012920.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>15888.0</td>\n",
       "      <td>15953.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-TRNT-1;Parent=rna-TRNT;Dbxref=GeneID:4...</td>\n",
       "      <td>exon-TRNT-1</td>\n",
       "      <td>rna-TRNT</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>TRNT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1698718</th>\n",
       "      <td>NC_012920.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>15956.0</td>\n",
       "      <td>16023.0</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=exon-TRNP-1;Parent=rna-TRNP;Dbxref=GeneID:4...</td>\n",
       "      <td>exon-TRNP-1</td>\n",
       "      <td>rna-TRNP</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>TRNP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1698719 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        nih_molecule_accession      source category    start     stop strand  \\\n",
       "0                 NC_000001.11  BestRefSeq     exon  17369.0  17436.0      -   \n",
       "1                 NC_000001.11  BestRefSeq     exon  17369.0  17391.0      -   \n",
       "2                 NC_000001.11  BestRefSeq     exon  17409.0  17431.0      -   \n",
       "3                 NC_000001.11      Gnomon     exon  29774.0  30667.0      +   \n",
       "4                 NC_000001.11      Gnomon     exon  30976.0  31093.0      +   \n",
       "...                        ...         ...      ...      ...      ...    ...   \n",
       "1698714            NC_012920.1      RefSeq     exon  14149.0  14673.0      -   \n",
       "1698715            NC_012920.1      RefSeq     exon  14674.0  14742.0      -   \n",
       "1698716            NC_012920.1      RefSeq     exon  14747.0  15887.0      +   \n",
       "1698717            NC_012920.1      RefSeq     exon  15888.0  15953.0      +   \n",
       "1698718            NC_012920.1      RefSeq     exon  15956.0  16023.0      -   \n",
       "\n",
       "                                               information  \\\n",
       "0        ID=exon-NR_106918.1-1;Parent=rna-NR_106918.1;D...   \n",
       "1        ID=exon-MIR6859-1-1;Parent=rna-MIR6859-1;Dbxre...   \n",
       "2        ID=exon-MIR6859-1-2-1;Parent=rna-MIR6859-1-2;D...   \n",
       "3        ID=exon-XR_007065314.1-1;Parent=rna-XR_0070653...   \n",
       "4        ID=exon-XR_007065314.1-2;Parent=rna-XR_0070653...   \n",
       "...                                                    ...   \n",
       "1698714  ID=exon-ND6-1;Parent=rna-ND6;Dbxref=GeneID:454...   \n",
       "1698715  ID=exon-TRNE-1;Parent=rna-TRNE;Dbxref=GeneID:4...   \n",
       "1698716  ID=exon-CYTB-1;Parent=rna-CYTB;Dbxref=GeneID:4...   \n",
       "1698717  ID=exon-TRNT-1;Parent=rna-TRNT;Dbxref=GeneID:4...   \n",
       "1698718  ID=exon-TRNP-1;Parent=rna-TRNP;Dbxref=GeneID:4...   \n",
       "\n",
       "                            ID         exon_parent     exon_gbkey  \\\n",
       "0           exon-NR_106918.1-1     rna-NR_106918.1  precursor_RNA   \n",
       "1             exon-MIR6859-1-1       rna-MIR6859-1          ncRNA   \n",
       "2           exon-MIR6859-1-2-1     rna-MIR6859-1-2          ncRNA   \n",
       "3        exon-XR_007065314.1-1  rna-XR_007065314.1          ncRNA   \n",
       "4        exon-XR_007065314.1-2  rna-XR_007065314.1          ncRNA   \n",
       "...                        ...                 ...            ...   \n",
       "1698714             exon-ND6-1             rna-ND6           mRNA   \n",
       "1698715            exon-TRNE-1            rna-TRNE           tRNA   \n",
       "1698716            exon-CYTB-1            rna-CYTB           mRNA   \n",
       "1698717            exon-TRNT-1            rna-TRNT           tRNA   \n",
       "1698718            exon-TRNP-1            rna-TRNP           tRNA   \n",
       "\n",
       "           exon_gene exon_transcript_id      predicted  \n",
       "0          MIR6859-1        NR_106918.1             no  \n",
       "1          MIR6859-1                NaN  not_available  \n",
       "2          MIR6859-1                NaN  not_available  \n",
       "3        MIR1302-2HG     XR_007065314.1            yes  \n",
       "4        MIR1302-2HG     XR_007065314.1            yes  \n",
       "...              ...                ...            ...  \n",
       "1698714          ND6                NaN  not_available  \n",
       "1698715         TRNE                NaN  not_available  \n",
       "1698716         CYTB                NaN  not_available  \n",
       "1698717         TRNT                NaN  not_available  \n",
       "1698718         TRNP                NaN  not_available  \n",
       "\n",
       "[1698719 rows x 13 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72252564",
   "metadata": {},
   "source": [
    "# Metrics (<= 25 introns lengths)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44e1f891",
   "metadata": {},
   "source": [
    "For the metrics that follow, we will concern ourselves only with exons for which a transcript ID is given (including predicted products)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f5928960",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Only want ones with transcripts.\n",
    "has_transcript = df[df['predicted'] != 'not_available'].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "30da4700",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>nih_molecule_accession</th>\n",
       "      <th>source</th>\n",
       "      <th>category</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>information</th>\n",
       "      <th>ID</th>\n",
       "      <th>exon_parent</th>\n",
       "      <th>exon_gbkey</th>\n",
       "      <th>exon_gene</th>\n",
       "      <th>exon_transcript_id</th>\n",
       "      <th>predicted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>17369.0</td>\n",
       "      <td>17436.0</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=exon-NR_106918.1-1;Parent=rna-NR_106918.1;D...</td>\n",
       "      <td>exon-NR_106918.1-1</td>\n",
       "      <td>rna-NR_106918.1</td>\n",
       "      <td>precursor_RNA</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>NR_106918.1</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>29774.0</td>\n",
       "      <td>30667.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-XR_007065314.1-1;Parent=rna-XR_0070653...</td>\n",
       "      <td>exon-XR_007065314.1-1</td>\n",
       "      <td>rna-XR_007065314.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>XR_007065314.1</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>30976.0</td>\n",
       "      <td>31093.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-XR_007065314.1-2;Parent=rna-XR_0070653...</td>\n",
       "      <td>exon-XR_007065314.1-2</td>\n",
       "      <td>rna-XR_007065314.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>XR_007065314.1</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>34168.0</td>\n",
       "      <td>35418.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-XR_007065314.1-3;Parent=rna-XR_0070653...</td>\n",
       "      <td>exon-XR_007065314.1-3</td>\n",
       "      <td>rna-XR_007065314.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>XR_007065314.1</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>30366.0</td>\n",
       "      <td>30503.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-NR_036051.1-1;Parent=rna-NR_036051.1;D...</td>\n",
       "      <td>exon-NR_036051.1-1</td>\n",
       "      <td>rna-NR_036051.1</td>\n",
       "      <td>precursor_RNA</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>NR_036051.1</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1698677</th>\n",
       "      <td>NC_000024.10</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>57192325.0</td>\n",
       "      <td>57192537.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-XM_017030055.2-1;Parent=rna-XM_0170300...</td>\n",
       "      <td>exon-XM_017030055.2-1</td>\n",
       "      <td>rna-XM_017030055.2</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>IL9R</td>\n",
       "      <td>XM_017030055.2</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1698678</th>\n",
       "      <td>NC_000024.10</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>57194043.0</td>\n",
       "      <td>57194127.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-XM_017030055.2-2;Parent=rna-XM_0170300...</td>\n",
       "      <td>exon-XM_017030055.2-2</td>\n",
       "      <td>rna-XM_017030055.2</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>IL9R</td>\n",
       "      <td>XM_017030055.2</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1698679</th>\n",
       "      <td>NC_000024.10</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>57196336.0</td>\n",
       "      <td>57197337.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-XM_017030055.2-3;Parent=rna-XM_0170300...</td>\n",
       "      <td>exon-XM_017030055.2-3</td>\n",
       "      <td>rna-XM_017030055.2</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>IL9R</td>\n",
       "      <td>XM_017030055.2</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1698680</th>\n",
       "      <td>NC_000024.10</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>57203182.0</td>\n",
       "      <td>57203350.0</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=exon-NR_138048.1-2-1;Parent=rna-NR_138048.1...</td>\n",
       "      <td>exon-NR_138048.1-2-1</td>\n",
       "      <td>rna-NR_138048.1-2</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>WASIR1</td>\n",
       "      <td>NR_138048.1</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1698681</th>\n",
       "      <td>NC_000024.10</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>57201084.0</td>\n",
       "      <td>57202145.0</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=exon-NR_138048.1-2-2;Parent=rna-NR_138048.1...</td>\n",
       "      <td>exon-NR_138048.1-2-2</td>\n",
       "      <td>rna-NR_138048.1-2</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>WASIR1</td>\n",
       "      <td>NR_138048.1</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1694667 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        nih_molecule_accession      source category       start        stop  \\\n",
       "0                 NC_000001.11  BestRefSeq     exon     17369.0     17436.0   \n",
       "3                 NC_000001.11      Gnomon     exon     29774.0     30667.0   \n",
       "4                 NC_000001.11      Gnomon     exon     30976.0     31093.0   \n",
       "5                 NC_000001.11      Gnomon     exon     34168.0     35418.0   \n",
       "6                 NC_000001.11  BestRefSeq     exon     30366.0     30503.0   \n",
       "...                        ...         ...      ...         ...         ...   \n",
       "1698677           NC_000024.10      Gnomon     exon  57192325.0  57192537.0   \n",
       "1698678           NC_000024.10      Gnomon     exon  57194043.0  57194127.0   \n",
       "1698679           NC_000024.10      Gnomon     exon  57196336.0  57197337.0   \n",
       "1698680           NC_000024.10  BestRefSeq     exon  57203182.0  57203350.0   \n",
       "1698681           NC_000024.10  BestRefSeq     exon  57201084.0  57202145.0   \n",
       "\n",
       "        strand                                        information  \\\n",
       "0            -  ID=exon-NR_106918.1-1;Parent=rna-NR_106918.1;D...   \n",
       "3            +  ID=exon-XR_007065314.1-1;Parent=rna-XR_0070653...   \n",
       "4            +  ID=exon-XR_007065314.1-2;Parent=rna-XR_0070653...   \n",
       "5            +  ID=exon-XR_007065314.1-3;Parent=rna-XR_0070653...   \n",
       "6            +  ID=exon-NR_036051.1-1;Parent=rna-NR_036051.1;D...   \n",
       "...        ...                                                ...   \n",
       "1698677      +  ID=exon-XM_017030055.2-1;Parent=rna-XM_0170300...   \n",
       "1698678      +  ID=exon-XM_017030055.2-2;Parent=rna-XM_0170300...   \n",
       "1698679      +  ID=exon-XM_017030055.2-3;Parent=rna-XM_0170300...   \n",
       "1698680      -  ID=exon-NR_138048.1-2-1;Parent=rna-NR_138048.1...   \n",
       "1698681      -  ID=exon-NR_138048.1-2-2;Parent=rna-NR_138048.1...   \n",
       "\n",
       "                            ID         exon_parent     exon_gbkey  \\\n",
       "0           exon-NR_106918.1-1     rna-NR_106918.1  precursor_RNA   \n",
       "3        exon-XR_007065314.1-1  rna-XR_007065314.1          ncRNA   \n",
       "4        exon-XR_007065314.1-2  rna-XR_007065314.1          ncRNA   \n",
       "5        exon-XR_007065314.1-3  rna-XR_007065314.1          ncRNA   \n",
       "6           exon-NR_036051.1-1     rna-NR_036051.1  precursor_RNA   \n",
       "...                        ...                 ...            ...   \n",
       "1698677  exon-XM_017030055.2-1  rna-XM_017030055.2           mRNA   \n",
       "1698678  exon-XM_017030055.2-2  rna-XM_017030055.2           mRNA   \n",
       "1698679  exon-XM_017030055.2-3  rna-XM_017030055.2           mRNA   \n",
       "1698680   exon-NR_138048.1-2-1   rna-NR_138048.1-2          ncRNA   \n",
       "1698681   exon-NR_138048.1-2-2   rna-NR_138048.1-2          ncRNA   \n",
       "\n",
       "           exon_gene exon_transcript_id predicted  \n",
       "0          MIR6859-1        NR_106918.1        no  \n",
       "3        MIR1302-2HG     XR_007065314.1       yes  \n",
       "4        MIR1302-2HG     XR_007065314.1       yes  \n",
       "5        MIR1302-2HG     XR_007065314.1       yes  \n",
       "6          MIR1302-2        NR_036051.1        no  \n",
       "...              ...                ...       ...  \n",
       "1698677         IL9R     XM_017030055.2       yes  \n",
       "1698678         IL9R     XM_017030055.2       yes  \n",
       "1698679         IL9R     XM_017030055.2       yes  \n",
       "1698680       WASIR1        NR_138048.1        no  \n",
       "1698681       WASIR1        NR_138048.1        no  \n",
       "\n",
       "[1694667 rows x 13 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "has_transcript"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9378c6a",
   "metadata": {},
   "source": [
    "Get exon information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "78ac5836",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Keep DataFrame.\n",
    "# Source: https://stackoverflow.com/a/42324086\n",
    "exon_counts = has_transcript.groupby(['exon_gene', 'exon_transcript_id'], as_index = False).size()\n",
    "exon_counts = exon_counts.rename(\n",
    "    columns = {\n",
    "        \"size\": \"n_exons\"\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f885db93",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon_gene</th>\n",
       "      <th>exon_transcript_id</th>\n",
       "      <th>n_exons</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>NM_130786.4</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>NR_015380.2</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>NM_001198818.2</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>NM_001198819.2</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>NM_001198820.2</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139214</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>XM_047417329.1</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139215</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>XM_047417333.1</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139216</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>XM_047417337.1</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139217</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>XM_047417341.1</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139218</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>XM_047417342.1</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>139219 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       exon_gene exon_transcript_id  n_exons\n",
       "0           A1BG        NM_130786.4        8\n",
       "1       A1BG-AS1        NR_015380.2        4\n",
       "2           A1CF     NM_001198818.2       14\n",
       "3           A1CF     NM_001198819.2       15\n",
       "4           A1CF     NM_001198820.2       14\n",
       "...          ...                ...      ...\n",
       "139214      ZZZ3     XM_047417329.1       14\n",
       "139215      ZZZ3     XM_047417333.1       13\n",
       "139216      ZZZ3     XM_047417337.1       14\n",
       "139217      ZZZ3     XM_047417341.1       13\n",
       "139218      ZZZ3     XM_047417342.1       13\n",
       "\n",
       "[139219 rows x 3 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon_counts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8051eb54",
   "metadata": {},
   "source": [
    "We can immediately identify the number of introns as well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1ebfd24b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add the intron counts.\n",
    "exon_counts['n_introns'] = exon_counts['n_exons']-1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "dbe198de",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon_gene</th>\n",
       "      <th>exon_transcript_id</th>\n",
       "      <th>n_exons</th>\n",
       "      <th>n_introns</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>NM_130786.4</td>\n",
       "      <td>8</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>NR_015380.2</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>NM_001198818.2</td>\n",
       "      <td>14</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>NM_001198819.2</td>\n",
       "      <td>15</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>NM_001198820.2</td>\n",
       "      <td>14</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139214</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>XM_047417329.1</td>\n",
       "      <td>14</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139215</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>XM_047417333.1</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139216</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>XM_047417337.1</td>\n",
       "      <td>14</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139217</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>XM_047417341.1</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139218</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>XM_047417342.1</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>139219 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       exon_gene exon_transcript_id  n_exons  n_introns\n",
       "0           A1BG        NM_130786.4        8          7\n",
       "1       A1BG-AS1        NR_015380.2        4          3\n",
       "2           A1CF     NM_001198818.2       14         13\n",
       "3           A1CF     NM_001198819.2       15         14\n",
       "4           A1CF     NM_001198820.2       14         13\n",
       "...          ...                ...      ...        ...\n",
       "139214      ZZZ3     XM_047417329.1       14         13\n",
       "139215      ZZZ3     XM_047417333.1       13         12\n",
       "139216      ZZZ3     XM_047417337.1       14         13\n",
       "139217      ZZZ3     XM_047417341.1       13         12\n",
       "139218      ZZZ3     XM_047417342.1       13         12\n",
       "\n",
       "[139219 rows x 4 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon_counts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54179827",
   "metadata": {},
   "source": [
    "Now we will explicitly calculate the possible lengths for transcripts with n_introns <= 25.  First, get the dataframe of such transcripts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c801bdd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1 <= n_introns <= 25.\n",
    "lte_25_introns = exon_counts[(exon_counts.n_introns >= 1) & (exon_counts.n_introns <= 25)].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b46c702e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon_gene</th>\n",
       "      <th>exon_transcript_id</th>\n",
       "      <th>n_exons</th>\n",
       "      <th>n_introns</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>NM_130786.4</td>\n",
       "      <td>8</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>NR_015380.2</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>NM_001198818.2</td>\n",
       "      <td>14</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>NM_001198819.2</td>\n",
       "      <td>15</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>NM_001198820.2</td>\n",
       "      <td>14</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139214</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>XM_047417329.1</td>\n",
       "      <td>14</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139215</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>XM_047417333.1</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139216</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>XM_047417337.1</td>\n",
       "      <td>14</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139217</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>XM_047417341.1</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139218</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>XM_047417342.1</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>122826 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       exon_gene exon_transcript_id  n_exons  n_introns\n",
       "0           A1BG        NM_130786.4        8          7\n",
       "1       A1BG-AS1        NR_015380.2        4          3\n",
       "2           A1CF     NM_001198818.2       14         13\n",
       "3           A1CF     NM_001198819.2       15         14\n",
       "4           A1CF     NM_001198820.2       14         13\n",
       "...          ...                ...      ...        ...\n",
       "139214      ZZZ3     XM_047417329.1       14         13\n",
       "139215      ZZZ3     XM_047417333.1       13         12\n",
       "139216      ZZZ3     XM_047417337.1       14         13\n",
       "139217      ZZZ3     XM_047417341.1       13         12\n",
       "139218      ZZZ3     XM_047417342.1       13         12\n",
       "\n",
       "[122826 rows x 4 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lte_25_introns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c813852",
   "metadata": {},
   "source": [
    "Subset to get the exon starts and stops."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3c9621fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Which transcripts?\n",
    "lte_25_introns_transcript_accessions = list(set(lte_25_introns['exon_transcript_id']))\n",
    "\n",
    "# Get exon positions.\n",
    "exon_filter = has_transcript['exon_transcript_id'].isin(lte_25_introns_transcript_accessions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b7451e4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "exon_positions = has_transcript[exon_filter].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "61d9e069",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>nih_molecule_accession</th>\n",
       "      <th>source</th>\n",
       "      <th>category</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>information</th>\n",
       "      <th>ID</th>\n",
       "      <th>exon_parent</th>\n",
       "      <th>exon_gbkey</th>\n",
       "      <th>exon_gene</th>\n",
       "      <th>exon_transcript_id</th>\n",
       "      <th>predicted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>29774.0</td>\n",
       "      <td>30667.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-XR_007065314.1-1;Parent=rna-XR_0070653...</td>\n",
       "      <td>exon-XR_007065314.1-1</td>\n",
       "      <td>rna-XR_007065314.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>XR_007065314.1</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>30976.0</td>\n",
       "      <td>31093.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-XR_007065314.1-2;Parent=rna-XR_0070653...</td>\n",
       "      <td>exon-XR_007065314.1-2</td>\n",
       "      <td>rna-XR_007065314.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>XR_007065314.1</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>34168.0</td>\n",
       "      <td>35418.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-XR_007065314.1-3;Parent=rna-XR_0070653...</td>\n",
       "      <td>exon-XR_007065314.1-3</td>\n",
       "      <td>rna-XR_007065314.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>XR_007065314.1</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>35721.0</td>\n",
       "      <td>36081.0</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=exon-NR_026818.1-1;Parent=rna-NR_026818.1;D...</td>\n",
       "      <td>exon-NR_026818.1-1</td>\n",
       "      <td>rna-NR_026818.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>NR_026818.1</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>35277.0</td>\n",
       "      <td>35481.0</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=exon-NR_026818.1-2;Parent=rna-NR_026818.1;D...</td>\n",
       "      <td>exon-NR_026818.1-2</td>\n",
       "      <td>rna-NR_026818.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>NR_026818.1</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1698677</th>\n",
       "      <td>NC_000024.10</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>57192325.0</td>\n",
       "      <td>57192537.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-XM_017030055.2-1;Parent=rna-XM_0170300...</td>\n",
       "      <td>exon-XM_017030055.2-1</td>\n",
       "      <td>rna-XM_017030055.2</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>IL9R</td>\n",
       "      <td>XM_017030055.2</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1698678</th>\n",
       "      <td>NC_000024.10</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>57194043.0</td>\n",
       "      <td>57194127.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-XM_017030055.2-2;Parent=rna-XM_0170300...</td>\n",
       "      <td>exon-XM_017030055.2-2</td>\n",
       "      <td>rna-XM_017030055.2</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>IL9R</td>\n",
       "      <td>XM_017030055.2</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1698679</th>\n",
       "      <td>NC_000024.10</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>57196336.0</td>\n",
       "      <td>57197337.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-XM_017030055.2-3;Parent=rna-XM_0170300...</td>\n",
       "      <td>exon-XM_017030055.2-3</td>\n",
       "      <td>rna-XM_017030055.2</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>IL9R</td>\n",
       "      <td>XM_017030055.2</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1698680</th>\n",
       "      <td>NC_000024.10</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>57203182.0</td>\n",
       "      <td>57203350.0</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=exon-NR_138048.1-2-1;Parent=rna-NR_138048.1...</td>\n",
       "      <td>exon-NR_138048.1-2-1</td>\n",
       "      <td>rna-NR_138048.1-2</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>WASIR1</td>\n",
       "      <td>NR_138048.1</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1698681</th>\n",
       "      <td>NC_000024.10</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>57201084.0</td>\n",
       "      <td>57202145.0</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=exon-NR_138048.1-2-2;Parent=rna-NR_138048.1...</td>\n",
       "      <td>exon-NR_138048.1-2-2</td>\n",
       "      <td>rna-NR_138048.1-2</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>WASIR1</td>\n",
       "      <td>NR_138048.1</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1251016 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        nih_molecule_accession      source category       start        stop  \\\n",
       "3                 NC_000001.11      Gnomon     exon     29774.0     30667.0   \n",
       "4                 NC_000001.11      Gnomon     exon     30976.0     31093.0   \n",
       "5                 NC_000001.11      Gnomon     exon     34168.0     35418.0   \n",
       "8                 NC_000001.11  BestRefSeq     exon     35721.0     36081.0   \n",
       "9                 NC_000001.11  BestRefSeq     exon     35277.0     35481.0   \n",
       "...                        ...         ...      ...         ...         ...   \n",
       "1698677           NC_000024.10      Gnomon     exon  57192325.0  57192537.0   \n",
       "1698678           NC_000024.10      Gnomon     exon  57194043.0  57194127.0   \n",
       "1698679           NC_000024.10      Gnomon     exon  57196336.0  57197337.0   \n",
       "1698680           NC_000024.10  BestRefSeq     exon  57203182.0  57203350.0   \n",
       "1698681           NC_000024.10  BestRefSeq     exon  57201084.0  57202145.0   \n",
       "\n",
       "        strand                                        information  \\\n",
       "3            +  ID=exon-XR_007065314.1-1;Parent=rna-XR_0070653...   \n",
       "4            +  ID=exon-XR_007065314.1-2;Parent=rna-XR_0070653...   \n",
       "5            +  ID=exon-XR_007065314.1-3;Parent=rna-XR_0070653...   \n",
       "8            -  ID=exon-NR_026818.1-1;Parent=rna-NR_026818.1;D...   \n",
       "9            -  ID=exon-NR_026818.1-2;Parent=rna-NR_026818.1;D...   \n",
       "...        ...                                                ...   \n",
       "1698677      +  ID=exon-XM_017030055.2-1;Parent=rna-XM_0170300...   \n",
       "1698678      +  ID=exon-XM_017030055.2-2;Parent=rna-XM_0170300...   \n",
       "1698679      +  ID=exon-XM_017030055.2-3;Parent=rna-XM_0170300...   \n",
       "1698680      -  ID=exon-NR_138048.1-2-1;Parent=rna-NR_138048.1...   \n",
       "1698681      -  ID=exon-NR_138048.1-2-2;Parent=rna-NR_138048.1...   \n",
       "\n",
       "                            ID         exon_parent exon_gbkey    exon_gene  \\\n",
       "3        exon-XR_007065314.1-1  rna-XR_007065314.1      ncRNA  MIR1302-2HG   \n",
       "4        exon-XR_007065314.1-2  rna-XR_007065314.1      ncRNA  MIR1302-2HG   \n",
       "5        exon-XR_007065314.1-3  rna-XR_007065314.1      ncRNA  MIR1302-2HG   \n",
       "8           exon-NR_026818.1-1     rna-NR_026818.1      ncRNA      FAM138A   \n",
       "9           exon-NR_026818.1-2     rna-NR_026818.1      ncRNA      FAM138A   \n",
       "...                        ...                 ...        ...          ...   \n",
       "1698677  exon-XM_017030055.2-1  rna-XM_017030055.2       mRNA         IL9R   \n",
       "1698678  exon-XM_017030055.2-2  rna-XM_017030055.2       mRNA         IL9R   \n",
       "1698679  exon-XM_017030055.2-3  rna-XM_017030055.2       mRNA         IL9R   \n",
       "1698680   exon-NR_138048.1-2-1   rna-NR_138048.1-2      ncRNA       WASIR1   \n",
       "1698681   exon-NR_138048.1-2-2   rna-NR_138048.1-2      ncRNA       WASIR1   \n",
       "\n",
       "        exon_transcript_id predicted  \n",
       "3           XR_007065314.1       yes  \n",
       "4           XR_007065314.1       yes  \n",
       "5           XR_007065314.1       yes  \n",
       "8              NR_026818.1        no  \n",
       "9              NR_026818.1        no  \n",
       "...                    ...       ...  \n",
       "1698677     XM_017030055.2       yes  \n",
       "1698678     XM_017030055.2       yes  \n",
       "1698679     XM_017030055.2       yes  \n",
       "1698680        NR_138048.1        no  \n",
       "1698681        NR_138048.1        no  \n",
       "\n",
       "[1251016 rows x 13 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon_positions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e48569f",
   "metadata": {},
   "source": [
    "Let's process the positive strand first."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fd9551e",
   "metadata": {},
   "source": [
    "## Positive strand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "1ea52d87",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Positive strand.\n",
    "positive_strand = exon_positions[exon_positions.strand == '+']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d30cd84e",
   "metadata": {},
   "source": [
    "Order by gene, then transcript, then start position."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "521e2a4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sort on gene, transcript, and start position.\n",
    "positive_strand = positive_strand.sort_values(by = ['exon_gene', 'exon_transcript_id', 'start'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ec3b7c72",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>nih_molecule_accession</th>\n",
       "      <th>source</th>\n",
       "      <th>category</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>information</th>\n",
       "      <th>ID</th>\n",
       "      <th>exon_parent</th>\n",
       "      <th>exon_gbkey</th>\n",
       "      <th>exon_gene</th>\n",
       "      <th>exon_transcript_id</th>\n",
       "      <th>predicted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1557862</th>\n",
       "      <td>NC_000019.10</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>58353379.0</td>\n",
       "      <td>58353474.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-NR_015380.2-2;Parent=rna-NR_015380.2;D...</td>\n",
       "      <td>exon-NR_015380.2-2</td>\n",
       "      <td>rna-NR_015380.2</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>NR_015380.2</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1557863</th>\n",
       "      <td>NC_000019.10</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>58353714.0</td>\n",
       "      <td>58353857.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-NR_015380.2-3;Parent=rna-NR_015380.2;D...</td>\n",
       "      <td>exon-NR_015380.2-3</td>\n",
       "      <td>rna-NR_015380.2</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>NR_015380.2</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1557864</th>\n",
       "      <td>NC_000019.10</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>58354369.0</td>\n",
       "      <td>58355183.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-NR_015380.2-4;Parent=rna-NR_015380.2;D...</td>\n",
       "      <td>exon-NR_015380.2-4</td>\n",
       "      <td>rna-NR_015380.2</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>NR_015380.2</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1060562</th>\n",
       "      <td>NC_000012.12</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>9065177.0</td>\n",
       "      <td>9065228.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-NR_026971.1-1;Parent=rna-NR_026971.1;D...</td>\n",
       "      <td>exon-NR_026971.1-1</td>\n",
       "      <td>rna-NR_026971.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>NR_026971.1</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1060563</th>\n",
       "      <td>NC_000012.12</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>9065826.0</td>\n",
       "      <td>9066060.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-NR_026971.1-2;Parent=rna-NR_026971.1;D...</td>\n",
       "      <td>exon-NR_026971.1-2</td>\n",
       "      <td>rna-NR_026971.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>NR_026971.1</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1060564</th>\n",
       "      <td>NC_000012.12</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>9066156.0</td>\n",
       "      <td>9068055.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-NR_026971.1-3;Parent=rna-NR_026971.1;D...</td>\n",
       "      <td>exon-NR_026971.1-3</td>\n",
       "      <td>rna-NR_026971.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>NR_026971.1</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1060558</th>\n",
       "      <td>NC_000012.12</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>9065177.0</td>\n",
       "      <td>9065228.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-NR_137424.1-1;Parent=rna-NR_137424.1;D...</td>\n",
       "      <td>exon-NR_137424.1-1</td>\n",
       "      <td>rna-NR_137424.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>NR_137424.1</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1060559</th>\n",
       "      <td>NC_000012.12</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>9065808.0</td>\n",
       "      <td>9068055.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-NR_137424.1-2;Parent=rna-NR_137424.1;D...</td>\n",
       "      <td>exon-NR_137424.1-2</td>\n",
       "      <td>rna-NR_137424.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>NR_137424.1</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1060560</th>\n",
       "      <td>NC_000012.12</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>9065177.0</td>\n",
       "      <td>9065228.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-NR_137425.1-1;Parent=rna-NR_137425.1;D...</td>\n",
       "      <td>exon-NR_137425.1-1</td>\n",
       "      <td>rna-NR_137425.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>NR_137425.1</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1060561</th>\n",
       "      <td>NC_000012.12</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>9065826.0</td>\n",
       "      <td>9068055.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-NR_137425.1-2;Parent=rna-NR_137425.1;D...</td>\n",
       "      <td>exon-NR_137425.1-2</td>\n",
       "      <td>rna-NR_137425.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>NR_137425.1</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1060016</th>\n",
       "      <td>NC_000012.12</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>8845004.0</td>\n",
       "      <td>8845174.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-NM_001282424.3-1;Parent=rna-NM_0012824...</td>\n",
       "      <td>exon-NM_001282424.3-1</td>\n",
       "      <td>rna-NM_001282424.3</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>NM_001282424.3</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1060017</th>\n",
       "      <td>NC_000012.12</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>8845442.0</td>\n",
       "      <td>8845502.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-NM_001282424.3-2;Parent=rna-NM_0012824...</td>\n",
       "      <td>exon-NM_001282424.3-2</td>\n",
       "      <td>rna-NM_001282424.3</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>NM_001282424.3</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1060018</th>\n",
       "      <td>NC_000012.12</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>8846077.0</td>\n",
       "      <td>8846222.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-NM_001282424.3-3;Parent=rna-NM_0012824...</td>\n",
       "      <td>exon-NM_001282424.3-3</td>\n",
       "      <td>rna-NM_001282424.3</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>NM_001282424.3</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1060019</th>\n",
       "      <td>NC_000012.12</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>8847549.0</td>\n",
       "      <td>8847698.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-NM_001282424.3-4;Parent=rna-NM_0012824...</td>\n",
       "      <td>exon-NM_001282424.3-4</td>\n",
       "      <td>rna-NM_001282424.3</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>NM_001282424.3</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1060020</th>\n",
       "      <td>NC_000012.12</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>8848720.0</td>\n",
       "      <td>8848914.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-NM_001282424.3-5;Parent=rna-NM_0012824...</td>\n",
       "      <td>exon-NM_001282424.3-5</td>\n",
       "      <td>rna-NM_001282424.3</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>NM_001282424.3</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1060021</th>\n",
       "      <td>NC_000012.12</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>8849669.0</td>\n",
       "      <td>8849759.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-NM_001282424.3-6;Parent=rna-NM_0012824...</td>\n",
       "      <td>exon-NM_001282424.3-6</td>\n",
       "      <td>rna-NM_001282424.3</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>NM_001282424.3</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1060022</th>\n",
       "      <td>NC_000012.12</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>8850160.0</td>\n",
       "      <td>8850274.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-NM_001282424.3-7;Parent=rna-NM_0012824...</td>\n",
       "      <td>exon-NM_001282424.3-7</td>\n",
       "      <td>rna-NM_001282424.3</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>NM_001282424.3</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1060023</th>\n",
       "      <td>NC_000012.12</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>8851784.0</td>\n",
       "      <td>8852012.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-NM_001282424.3-8;Parent=rna-NM_0012824...</td>\n",
       "      <td>exon-NM_001282424.3-8</td>\n",
       "      <td>rna-NM_001282424.3</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>NM_001282424.3</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1060024</th>\n",
       "      <td>NC_000012.12</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>8852210.0</td>\n",
       "      <td>8852336.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-NM_001282424.3-9;Parent=rna-NM_0012824...</td>\n",
       "      <td>exon-NM_001282424.3-9</td>\n",
       "      <td>rna-NM_001282424.3</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>NM_001282424.3</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        nih_molecule_accession      source category       start        stop  \\\n",
       "1557862           NC_000019.10  BestRefSeq     exon  58353379.0  58353474.0   \n",
       "1557863           NC_000019.10  BestRefSeq     exon  58353714.0  58353857.0   \n",
       "1557864           NC_000019.10  BestRefSeq     exon  58354369.0  58355183.0   \n",
       "1060562           NC_000012.12  BestRefSeq     exon   9065177.0   9065228.0   \n",
       "1060563           NC_000012.12  BestRefSeq     exon   9065826.0   9066060.0   \n",
       "1060564           NC_000012.12  BestRefSeq     exon   9066156.0   9068055.0   \n",
       "1060558           NC_000012.12  BestRefSeq     exon   9065177.0   9065228.0   \n",
       "1060559           NC_000012.12  BestRefSeq     exon   9065808.0   9068055.0   \n",
       "1060560           NC_000012.12  BestRefSeq     exon   9065177.0   9065228.0   \n",
       "1060561           NC_000012.12  BestRefSeq     exon   9065826.0   9068055.0   \n",
       "1060016           NC_000012.12  BestRefSeq     exon   8845004.0   8845174.0   \n",
       "1060017           NC_000012.12  BestRefSeq     exon   8845442.0   8845502.0   \n",
       "1060018           NC_000012.12  BestRefSeq     exon   8846077.0   8846222.0   \n",
       "1060019           NC_000012.12  BestRefSeq     exon   8847549.0   8847698.0   \n",
       "1060020           NC_000012.12  BestRefSeq     exon   8848720.0   8848914.0   \n",
       "1060021           NC_000012.12  BestRefSeq     exon   8849669.0   8849759.0   \n",
       "1060022           NC_000012.12  BestRefSeq     exon   8850160.0   8850274.0   \n",
       "1060023           NC_000012.12  BestRefSeq     exon   8851784.0   8852012.0   \n",
       "1060024           NC_000012.12  BestRefSeq     exon   8852210.0   8852336.0   \n",
       "\n",
       "        strand                                        information  \\\n",
       "1557862      +  ID=exon-NR_015380.2-2;Parent=rna-NR_015380.2;D...   \n",
       "1557863      +  ID=exon-NR_015380.2-3;Parent=rna-NR_015380.2;D...   \n",
       "1557864      +  ID=exon-NR_015380.2-4;Parent=rna-NR_015380.2;D...   \n",
       "1060562      +  ID=exon-NR_026971.1-1;Parent=rna-NR_026971.1;D...   \n",
       "1060563      +  ID=exon-NR_026971.1-2;Parent=rna-NR_026971.1;D...   \n",
       "1060564      +  ID=exon-NR_026971.1-3;Parent=rna-NR_026971.1;D...   \n",
       "1060558      +  ID=exon-NR_137424.1-1;Parent=rna-NR_137424.1;D...   \n",
       "1060559      +  ID=exon-NR_137424.1-2;Parent=rna-NR_137424.1;D...   \n",
       "1060560      +  ID=exon-NR_137425.1-1;Parent=rna-NR_137425.1;D...   \n",
       "1060561      +  ID=exon-NR_137425.1-2;Parent=rna-NR_137425.1;D...   \n",
       "1060016      +  ID=exon-NM_001282424.3-1;Parent=rna-NM_0012824...   \n",
       "1060017      +  ID=exon-NM_001282424.3-2;Parent=rna-NM_0012824...   \n",
       "1060018      +  ID=exon-NM_001282424.3-3;Parent=rna-NM_0012824...   \n",
       "1060019      +  ID=exon-NM_001282424.3-4;Parent=rna-NM_0012824...   \n",
       "1060020      +  ID=exon-NM_001282424.3-5;Parent=rna-NM_0012824...   \n",
       "1060021      +  ID=exon-NM_001282424.3-6;Parent=rna-NM_0012824...   \n",
       "1060022      +  ID=exon-NM_001282424.3-7;Parent=rna-NM_0012824...   \n",
       "1060023      +  ID=exon-NM_001282424.3-8;Parent=rna-NM_0012824...   \n",
       "1060024      +  ID=exon-NM_001282424.3-9;Parent=rna-NM_0012824...   \n",
       "\n",
       "                            ID         exon_parent exon_gbkey exon_gene  \\\n",
       "1557862     exon-NR_015380.2-2     rna-NR_015380.2      ncRNA  A1BG-AS1   \n",
       "1557863     exon-NR_015380.2-3     rna-NR_015380.2      ncRNA  A1BG-AS1   \n",
       "1557864     exon-NR_015380.2-4     rna-NR_015380.2      ncRNA  A1BG-AS1   \n",
       "1060562     exon-NR_026971.1-1     rna-NR_026971.1      ncRNA   A2M-AS1   \n",
       "1060563     exon-NR_026971.1-2     rna-NR_026971.1      ncRNA   A2M-AS1   \n",
       "1060564     exon-NR_026971.1-3     rna-NR_026971.1      ncRNA   A2M-AS1   \n",
       "1060558     exon-NR_137424.1-1     rna-NR_137424.1      ncRNA   A2M-AS1   \n",
       "1060559     exon-NR_137424.1-2     rna-NR_137424.1      ncRNA   A2M-AS1   \n",
       "1060560     exon-NR_137425.1-1     rna-NR_137425.1      ncRNA   A2M-AS1   \n",
       "1060561     exon-NR_137425.1-2     rna-NR_137425.1      ncRNA   A2M-AS1   \n",
       "1060016  exon-NM_001282424.3-1  rna-NM_001282424.3       mRNA     A2ML1   \n",
       "1060017  exon-NM_001282424.3-2  rna-NM_001282424.3       mRNA     A2ML1   \n",
       "1060018  exon-NM_001282424.3-3  rna-NM_001282424.3       mRNA     A2ML1   \n",
       "1060019  exon-NM_001282424.3-4  rna-NM_001282424.3       mRNA     A2ML1   \n",
       "1060020  exon-NM_001282424.3-5  rna-NM_001282424.3       mRNA     A2ML1   \n",
       "1060021  exon-NM_001282424.3-6  rna-NM_001282424.3       mRNA     A2ML1   \n",
       "1060022  exon-NM_001282424.3-7  rna-NM_001282424.3       mRNA     A2ML1   \n",
       "1060023  exon-NM_001282424.3-8  rna-NM_001282424.3       mRNA     A2ML1   \n",
       "1060024  exon-NM_001282424.3-9  rna-NM_001282424.3       mRNA     A2ML1   \n",
       "\n",
       "        exon_transcript_id predicted  \n",
       "1557862        NR_015380.2        no  \n",
       "1557863        NR_015380.2        no  \n",
       "1557864        NR_015380.2        no  \n",
       "1060562        NR_026971.1        no  \n",
       "1060563        NR_026971.1        no  \n",
       "1060564        NR_026971.1        no  \n",
       "1060558        NR_137424.1        no  \n",
       "1060559        NR_137424.1        no  \n",
       "1060560        NR_137425.1        no  \n",
       "1060561        NR_137425.1        no  \n",
       "1060016     NM_001282424.3        no  \n",
       "1060017     NM_001282424.3        no  \n",
       "1060018     NM_001282424.3        no  \n",
       "1060019     NM_001282424.3        no  \n",
       "1060020     NM_001282424.3        no  \n",
       "1060021     NM_001282424.3        no  \n",
       "1060022     NM_001282424.3        no  \n",
       "1060023     NM_001282424.3        no  \n",
       "1060024     NM_001282424.3        no  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "positive_strand[1:20]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb439570",
   "metadata": {},
   "source": [
    "We can avoid expensive grouping and subtraction operations by making three lists (starts,stops, and transcript ID), and subtracting the lists.  We then exclude meaningless subtractions (e.g. between different transcripts)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "24eae979",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Starts, stops, and IDs.\n",
    "starts = list(positive_strand.start)[1:]\n",
    "stops = list(positive_strand.stop)[:-1]\n",
    "etids_starts = list(positive_strand['exon_transcript_id'])[1:]\n",
    "etids_stops = list(positive_strand['exon_transcript_id'])[:-1]\n",
    "\n",
    "# The intron lengths.\n",
    "intron_lengths = {}\n",
    "\n",
    "# Double set to ensure we get all IDs.\n",
    "for tid in set(set(etids_starts).union(set(etids_stops))):\n",
    "    intron_lengths[tid] = []\n",
    "\n",
    "# Subtract.\n",
    "for i in range(1, len(starts)):\n",
    "# for i in range(1, 20):\n",
    "    \n",
    "    # Exploit the ordered nature of the data.\n",
    "    start_tid = etids_starts[i]\n",
    "    stop_tid = etids_stops[i]\n",
    "    \n",
    "    if start_tid == stop_tid:\n",
    "        \n",
    "        # TODO: fix data type for start and stop columns\n",
    "        # earlier in analysis because we are doing\n",
    "        # int typecasting here.\n",
    "        intron_lengths[start_tid].append(int(starts[i] - stops[i]) - 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "eb23ea57",
   "metadata": {},
   "outputs": [],
   "source": [
    "# intron_lengths"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d8f1090",
   "metadata": {},
   "source": [
    "Use even simpler logic to find the exon lengths."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "8854838d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make an exon length column.\n",
    "positive_strand['exon_length'] = positive_strand.stop - positive_strand.start + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "d47b15b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>nih_molecule_accession</th>\n",
       "      <th>source</th>\n",
       "      <th>category</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>information</th>\n",
       "      <th>ID</th>\n",
       "      <th>exon_parent</th>\n",
       "      <th>exon_gbkey</th>\n",
       "      <th>exon_gene</th>\n",
       "      <th>exon_transcript_id</th>\n",
       "      <th>predicted</th>\n",
       "      <th>exon_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1557861</th>\n",
       "      <td>NC_000019.10</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>58351970.0</td>\n",
       "      <td>58353044.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-NR_015380.2-1;Parent=rna-NR_015380.2;D...</td>\n",
       "      <td>exon-NR_015380.2-1</td>\n",
       "      <td>rna-NR_015380.2</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>NR_015380.2</td>\n",
       "      <td>no</td>\n",
       "      <td>1075.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1557862</th>\n",
       "      <td>NC_000019.10</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>58353379.0</td>\n",
       "      <td>58353474.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-NR_015380.2-2;Parent=rna-NR_015380.2;D...</td>\n",
       "      <td>exon-NR_015380.2-2</td>\n",
       "      <td>rna-NR_015380.2</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>NR_015380.2</td>\n",
       "      <td>no</td>\n",
       "      <td>96.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1557863</th>\n",
       "      <td>NC_000019.10</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>58353714.0</td>\n",
       "      <td>58353857.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-NR_015380.2-3;Parent=rna-NR_015380.2;D...</td>\n",
       "      <td>exon-NR_015380.2-3</td>\n",
       "      <td>rna-NR_015380.2</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>NR_015380.2</td>\n",
       "      <td>no</td>\n",
       "      <td>144.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1557864</th>\n",
       "      <td>NC_000019.10</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>58354369.0</td>\n",
       "      <td>58355183.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-NR_015380.2-4;Parent=rna-NR_015380.2;D...</td>\n",
       "      <td>exon-NR_015380.2-4</td>\n",
       "      <td>rna-NR_015380.2</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>NR_015380.2</td>\n",
       "      <td>no</td>\n",
       "      <td>815.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1060562</th>\n",
       "      <td>NC_000012.12</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>9065177.0</td>\n",
       "      <td>9065228.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-NR_026971.1-1;Parent=rna-NR_026971.1;D...</td>\n",
       "      <td>exon-NR_026971.1-1</td>\n",
       "      <td>rna-NR_026971.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>NR_026971.1</td>\n",
       "      <td>no</td>\n",
       "      <td>52.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>731296</th>\n",
       "      <td>NC_000007.14</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>143381557.0</td>\n",
       "      <td>143381779.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-XM_047420818.1-2;Parent=rna-XM_0474208...</td>\n",
       "      <td>exon-XM_047420818.1-2</td>\n",
       "      <td>rna-XM_047420818.1</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>ZYX</td>\n",
       "      <td>XM_047420818.1</td>\n",
       "      <td>yes</td>\n",
       "      <td>223.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>731297</th>\n",
       "      <td>NC_000007.14</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>143382248.0</td>\n",
       "      <td>143382447.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-XM_047420818.1-3;Parent=rna-XM_0474208...</td>\n",
       "      <td>exon-XM_047420818.1-3</td>\n",
       "      <td>rna-XM_047420818.1</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>ZYX</td>\n",
       "      <td>XM_047420818.1</td>\n",
       "      <td>yes</td>\n",
       "      <td>200.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>731298</th>\n",
       "      <td>NC_000007.14</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>143382593.0</td>\n",
       "      <td>143382685.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-XM_047420818.1-4;Parent=rna-XM_0474208...</td>\n",
       "      <td>exon-XM_047420818.1-4</td>\n",
       "      <td>rna-XM_047420818.1</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>ZYX</td>\n",
       "      <td>XM_047420818.1</td>\n",
       "      <td>yes</td>\n",
       "      <td>93.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>731299</th>\n",
       "      <td>NC_000007.14</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>143382801.0</td>\n",
       "      <td>143383322.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-XM_047420818.1-5;Parent=rna-XM_0474208...</td>\n",
       "      <td>exon-XM_047420818.1-5</td>\n",
       "      <td>rna-XM_047420818.1</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>ZYX</td>\n",
       "      <td>XM_047420818.1</td>\n",
       "      <td>yes</td>\n",
       "      <td>522.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>731300</th>\n",
       "      <td>NC_000007.14</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>143384090.0</td>\n",
       "      <td>143385385.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-XM_047420818.1-6;Parent=rna-XM_0474208...</td>\n",
       "      <td>exon-XM_047420818.1-6</td>\n",
       "      <td>rna-XM_047420818.1</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>ZYX</td>\n",
       "      <td>XM_047420818.1</td>\n",
       "      <td>yes</td>\n",
       "      <td>1296.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>635200 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        nih_molecule_accession      source category        start         stop  \\\n",
       "1557861           NC_000019.10  BestRefSeq     exon   58351970.0   58353044.0   \n",
       "1557862           NC_000019.10  BestRefSeq     exon   58353379.0   58353474.0   \n",
       "1557863           NC_000019.10  BestRefSeq     exon   58353714.0   58353857.0   \n",
       "1557864           NC_000019.10  BestRefSeq     exon   58354369.0   58355183.0   \n",
       "1060562           NC_000012.12  BestRefSeq     exon    9065177.0    9065228.0   \n",
       "...                        ...         ...      ...          ...          ...   \n",
       "731296            NC_000007.14      Gnomon     exon  143381557.0  143381779.0   \n",
       "731297            NC_000007.14      Gnomon     exon  143382248.0  143382447.0   \n",
       "731298            NC_000007.14      Gnomon     exon  143382593.0  143382685.0   \n",
       "731299            NC_000007.14      Gnomon     exon  143382801.0  143383322.0   \n",
       "731300            NC_000007.14      Gnomon     exon  143384090.0  143385385.0   \n",
       "\n",
       "        strand                                        information  \\\n",
       "1557861      +  ID=exon-NR_015380.2-1;Parent=rna-NR_015380.2;D...   \n",
       "1557862      +  ID=exon-NR_015380.2-2;Parent=rna-NR_015380.2;D...   \n",
       "1557863      +  ID=exon-NR_015380.2-3;Parent=rna-NR_015380.2;D...   \n",
       "1557864      +  ID=exon-NR_015380.2-4;Parent=rna-NR_015380.2;D...   \n",
       "1060562      +  ID=exon-NR_026971.1-1;Parent=rna-NR_026971.1;D...   \n",
       "...        ...                                                ...   \n",
       "731296       +  ID=exon-XM_047420818.1-2;Parent=rna-XM_0474208...   \n",
       "731297       +  ID=exon-XM_047420818.1-3;Parent=rna-XM_0474208...   \n",
       "731298       +  ID=exon-XM_047420818.1-4;Parent=rna-XM_0474208...   \n",
       "731299       +  ID=exon-XM_047420818.1-5;Parent=rna-XM_0474208...   \n",
       "731300       +  ID=exon-XM_047420818.1-6;Parent=rna-XM_0474208...   \n",
       "\n",
       "                            ID         exon_parent exon_gbkey exon_gene  \\\n",
       "1557861     exon-NR_015380.2-1     rna-NR_015380.2      ncRNA  A1BG-AS1   \n",
       "1557862     exon-NR_015380.2-2     rna-NR_015380.2      ncRNA  A1BG-AS1   \n",
       "1557863     exon-NR_015380.2-3     rna-NR_015380.2      ncRNA  A1BG-AS1   \n",
       "1557864     exon-NR_015380.2-4     rna-NR_015380.2      ncRNA  A1BG-AS1   \n",
       "1060562     exon-NR_026971.1-1     rna-NR_026971.1      ncRNA   A2M-AS1   \n",
       "...                        ...                 ...        ...       ...   \n",
       "731296   exon-XM_047420818.1-2  rna-XM_047420818.1       mRNA       ZYX   \n",
       "731297   exon-XM_047420818.1-3  rna-XM_047420818.1       mRNA       ZYX   \n",
       "731298   exon-XM_047420818.1-4  rna-XM_047420818.1       mRNA       ZYX   \n",
       "731299   exon-XM_047420818.1-5  rna-XM_047420818.1       mRNA       ZYX   \n",
       "731300   exon-XM_047420818.1-6  rna-XM_047420818.1       mRNA       ZYX   \n",
       "\n",
       "        exon_transcript_id predicted  exon_length  \n",
       "1557861        NR_015380.2        no       1075.0  \n",
       "1557862        NR_015380.2        no         96.0  \n",
       "1557863        NR_015380.2        no        144.0  \n",
       "1557864        NR_015380.2        no        815.0  \n",
       "1060562        NR_026971.1        no         52.0  \n",
       "...                    ...       ...          ...  \n",
       "731296      XM_047420818.1       yes        223.0  \n",
       "731297      XM_047420818.1       yes        200.0  \n",
       "731298      XM_047420818.1       yes         93.0  \n",
       "731299      XM_047420818.1       yes        522.0  \n",
       "731300      XM_047420818.1       yes       1296.0  \n",
       "\n",
       "[635200 rows x 14 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "positive_strand"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f1cf13d",
   "metadata": {},
   "source": [
    "Get the exon lengths."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "db419eb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get all exon lengths.\n",
    "exons = list(positive_strand.exon_length)\n",
    "\n",
    "# Get all transcripts.\n",
    "exon_transcripts = list(positive_strand.exon_transcript_id)\n",
    "\n",
    "# The exon lengths.\n",
    "exon_lengths = {}\n",
    "\n",
    "# Go position by position and see where\n",
    "# the exons belong.\n",
    "for i in range(0, len(exon_transcripts)):\n",
    "    \n",
    "    # Get the transcript first.\n",
    "    tscript = exon_transcripts[i]\n",
    "    \n",
    "    # In the dictionary already?\n",
    "    if tscript not in exon_lengths:\n",
    "        exon_lengths[tscript] = []\n",
    "    \n",
    "    # Now append.\n",
    "    exon_lengths[tscript].append(int(exons[i]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "cf7052f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# exon_lengths"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f443bccf",
   "metadata": {},
   "source": [
    "Check to see if the lengths of the intron list and the exon list match."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "4274601a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(intron_lengths.keys()) == len(exon_lengths.keys())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b1ed0aa",
   "metadata": {},
   "source": [
    "We can define minimum and maximum possible transcript lengths easily."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "cebe1b38",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Maximum possible transcript lengths.\n",
    "transcript_length_extrema = {}\n",
    "\n",
    "for k, v in exon_lengths.items():\n",
    "    transcript_length_extrema[k] = {\n",
    "        'minimum_possible_transcript_length': sum(v),\n",
    "        'maximum_possible_transcript_length': sum(v) + sum(intron_lengths[k])\n",
    "    }"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c21cf19",
   "metadata": {},
   "source": [
    "Now go transcript by transcript and find all possible transcript lengths, saving the information to file (expensive).  Start by getting all combinations of 1 to 25 introns.\n",
    "\n",
    "TODO: split into files 1_1, 1_2, 1_3, ..., 1_25 and load based on the transcript list we are actually working with."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cec9751a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# All intronic combinations up to 25.\n",
    "lte_25_intron_combinations = {}\n",
    "\n",
    "# The combinations.\n",
    "with open('/media/apollo/Samsung_T5/transfer/mayur/1_25_combinations.tsv', 'r') as f:\n",
    "    \n",
    "    combos = f.readlines()[1:]\n",
    "    \n",
    "    # Initialize the dictionary.\n",
    "    for i in range(1, 26):\n",
    "        \n",
    "        lte_25_intron_combinations[i] = {}\n",
    "        \n",
    "        for j in range(1, i + 1):\n",
    "            \n",
    "            lte_25_intron_combinations[i][j] = []\n",
    "    \n",
    "    # Now go through the combinations\n",
    "    # and populate the dictionary.\n",
    "    for combo in combos:\n",
    "        \n",
    "        # Strip and split.\n",
    "        stripped_and_split = combo.strip().split('\\t')\n",
    "        \n",
    "        # Store it.\n",
    "        lte_25_intron_combinations[int(stripped_and_split[0])][int(stripped_and_split[1])].append(tuple(stripped_and_split[2:]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c58a877",
   "metadata": {},
   "outputs": [],
   "source": [
    "# lte_25_intron_combinations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "055d2489",
   "metadata": {},
   "source": [
    "The transcripts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67ec3c4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# The file to write to.\n",
    "with open('intermediate_states.tsv', 'w') as f:\n",
    "    \n",
    "    # Write the header.\n",
    "    f.write('\\t'.join(['transcript_id', 'minimum_transcript_length', 'maximum_transcript_length', 'n_exons_in_transcript', 'n_introns_in_transcript', 'index', 'choose_number', 'combination', 'intermediate_transcript_length']) + '\\n')\n",
    "\n",
    "    # Each transcript.\n",
    "    for transcript_id in list(intron_lengths.keys())[0:5]:\n",
    "\n",
    "        # Get the extrema of the transcript length.        \n",
    "        minimum_transcript_length = transcript_length_extrema[transcript_id]['minimum_possible_transcript_length']\n",
    "        maximum_transcript_length = transcript_length_extrema[transcript_id]['maximum_possible_transcript_length']\n",
    "\n",
    "        # Get the number of introns and exons in this transcript.\n",
    "        n_introns_in_transcript = len(intron_lengths[transcript_id])\n",
    "        n_exons_in_transcript = len(exon_lengths[transcript_id])\n",
    "        \n",
    "        # Write the \"no selection\" line.\n",
    "        \n",
    "        # TODO: put a 0 selection in the 1_25 file \n",
    "        # so we can use the file instead?\n",
    "        f.write('\\t'.join([\n",
    "            transcript_id, \n",
    "            str(minimum_transcript_length), \n",
    "            str(maximum_transcript_length), \n",
    "            str(n_exons_in_transcript),\n",
    "            str(n_introns_in_transcript),\n",
    "            '0',\n",
    "            '0', \n",
    "            '0', \n",
    "            str(minimum_transcript_length)\n",
    "        ]) + '\\n')\n",
    "\n",
    "        # Get the possible combinations for this number\n",
    "        # of introns.\n",
    "        combos = lte_25_intron_combinations[n_introns_in_transcript]\n",
    "\n",
    "        # We will create an index so that we can\n",
    "        # more easily track combinations.\n",
    "        index = 1\n",
    "\n",
    "        # Now iterate over the combinations.\n",
    "        for choose_number in combos.keys():\n",
    "            for selections in combos[choose_number]:\n",
    "\n",
    "                # Create a variable to hold the length.\n",
    "                intermediate_transcript_length = minimum_transcript_length\n",
    "\n",
    "                # Get exactly which introns we're selecting.\n",
    "                for intron_number in selections[0].split(','):\n",
    "                    \n",
    "                    # intron_number adjusted down because of 0-indexing\n",
    "                    # in python.\n",
    "\n",
    "                    # Select the intron and add to the length.\n",
    "                    intermediate_transcript_length += intron_lengths[transcript_id][int(intron_number) - 1]\n",
    "                \n",
    "                # Write to file.\n",
    "                f.write('\\t'.join([\n",
    "                    transcript_id, \n",
    "                    str(minimum_transcript_length), \n",
    "                    str(maximum_transcript_length), \n",
    "                    str(n_exons_in_transcript),\n",
    "                    str(n_introns_in_transcript),\n",
    "                    str(index),\n",
    "                    str(choose_number), \n",
    "                    selections[0], \n",
    "                    str(intermediate_transcript_length)\n",
    "                ]) + '\\n')\n",
    "                \n",
    "                # Increase the index.\n",
    "                index += 1  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7350f667",
   "metadata": {},
   "source": [
    "## Negative strand"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76e29785",
   "metadata": {},
   "source": [
    "We use largely the same logic as above, with the exception of the order for the start and stop positions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2997240",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Positive strand.\n",
    "negative_strand = exon_positions[exon_positions.strand == '-']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b589bc0a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "negative_strand"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b7590e3",
   "metadata": {},
   "source": [
    "Order by gene, then transcript, then stop position."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07c87a73",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sort on gene, transcript, and start position.\n",
    "negative_strand = negative_strand.sort_values(ascending = [True, True, False], by = ['exon_gene', 'exon_transcript_id', 'stop'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df100dcb",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "negative_strand[1:30]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "062de7ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Starts, stops, and IDs.\n",
    "starts = list(negative_strand.stop)[:-1]\n",
    "stops = list(negative_strand.start)[1:]\n",
    "etids_starts = list(negative_strand['exon_transcript_id'])[:-1]\n",
    "etids_stops = list(negative_strand['exon_transcript_id'])[1:]\n",
    "\n",
    "# The intron lengths.\n",
    "intron_lengths = {}\n",
    "\n",
    "# Double set to ensure we get all IDs.\n",
    "for tid in set(set(etids_starts).union(set(etids_stops))):\n",
    "    intron_lengths[tid] = []\n",
    "\n",
    "# Subtract.\n",
    "for i in range(1, len(starts)):\n",
    "# for i in range(1, 20):\n",
    "    \n",
    "    # Exploit the ordered nature of the data.\n",
    "    start_tid = etids_starts[i]\n",
    "    stop_tid = etids_stops[i]\n",
    "    \n",
    "    if start_tid == stop_tid:\n",
    "        \n",
    "        # TODO: fix data type for start and stop columns\n",
    "        # earlier in analysis because we are doing\n",
    "        # int typecasting here.\n",
    "        intron_lengths[start_tid].append(int(starts[i] - stops[i]) - 1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e163261d",
   "metadata": {},
   "source": [
    "Use even simpler logic to find the exon lengths."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f1d0f14",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make an exon length column.\n",
    "negative_strand['exon_length'] = negative_strand.stop - negative_strand.start + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e99b265c",
   "metadata": {},
   "outputs": [],
   "source": [
    "negative_strand"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40c2634c",
   "metadata": {},
   "source": [
    "Get the exon lengths."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc3eded4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get all exon lengths.\n",
    "exons = list(negative_strand.exon_length)\n",
    "\n",
    "# Get all transcripts.\n",
    "exon_transcripts = list(negative_strand.exon_transcript_id)\n",
    "\n",
    "# The exon lengths.\n",
    "exon_lengths = {}\n",
    "\n",
    "# Go position by position and see where\n",
    "# the exons belong.\n",
    "for i in range(0, len(exon_transcripts)):\n",
    "    \n",
    "    # Get the transcript first.\n",
    "    tscript = exon_transcripts[i]\n",
    "    \n",
    "    # In the dictionary already?\n",
    "    if tscript not in exon_lengths:\n",
    "        exon_lengths[tscript] = []\n",
    "    \n",
    "    # Now append.\n",
    "    exon_lengths[tscript].append(int(exons[i]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "424a633d",
   "metadata": {},
   "outputs": [],
   "source": [
    "exon_lengths"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9fb55563",
   "metadata": {},
   "source": [
    "Check to see if the lengths of the intron list and the exon list match."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00ef102a",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(intron_lengths.keys()) == len(exon_lengths.keys())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2714b0e0",
   "metadata": {},
   "source": [
    "We can define minimum and maximum possible transcript lengths easily."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d01c414",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Maximum possible transcript lengths.\n",
    "transcript_length_extrema = {}\n",
    "\n",
    "for k, v in exon_lengths.items():\n",
    "    transcript_length_extrema[k] = {\n",
    "        'minimum_possible_transcript_length': sum(v),\n",
    "        'maximum_possible_transcript_length': sum(v) + sum(intron_lengths[k])\n",
    "    }"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "822ba3f3",
   "metadata": {},
   "source": [
    "Now go transcript by transcript and find all possible transcript lengths, saving the information to file (expensive).  Start by getting all combinations of 1 to 25 introns.\n",
    "\n",
    "TODO: split into files 1_1, 1_2, 1_3, ..., 1_25 and load based on the transcript list we are actually working with."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ac20bff",
   "metadata": {},
   "source": [
    "The transcripts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c834294",
   "metadata": {},
   "outputs": [],
   "source": [
    "# The file to write to.\n",
    "with open('intermediate_states.tsv', 'a') as f:\n",
    "    \n",
    "    # Write the header.\n",
    "    f.write('\\t'.join(['transcript_id', 'minimum_transcript_length', 'maximum_transcript_length', 'n_exons_in_transcript', 'n_introns_in_transcript', 'index', 'choose_number', 'combination', 'intermediate_transcript_length']) + '\\n')\n",
    "\n",
    "    # Each transcript.\n",
    "    for transcript_id in list(intron_lengths.keys())[0:5]:\n",
    "\n",
    "        # Get the extrema of the transcript length.        \n",
    "        minimum_transcript_length = transcript_length_extrema[transcript_id]['minimum_possible_transcript_length']\n",
    "        maximum_transcript_length = transcript_length_extrema[transcript_id]['maximum_possible_transcript_length']\n",
    "\n",
    "        # Get the number of introns and exons in this transcript.\n",
    "        n_introns_in_transcript = len(intron_lengths[transcript_id])\n",
    "        n_exons_in_transcript = len(exon_lengths[transcript_id])\n",
    "        \n",
    "        # Write the \"no selection\" line.\n",
    "        \n",
    "        # TODO: put a 0 selection in the 1_25 file \n",
    "        # so we can use the file instead?\n",
    "        f.write('\\t'.join([\n",
    "            transcript_id, \n",
    "            str(minimum_transcript_length), \n",
    "            str(maximum_transcript_length), \n",
    "            str(n_exons_in_transcript),\n",
    "            str(n_introns_in_transcript),\n",
    "            '0',\n",
    "            '0', \n",
    "            '0', \n",
    "            str(minimum_transcript_length)\n",
    "        ]) + '\\n')\n",
    "\n",
    "        # Get the possible combinations for this number\n",
    "        # of introns.\n",
    "        combos = lte_25_intron_combinations[n_introns_in_transcript]\n",
    "\n",
    "        # We will create an index so that we can\n",
    "        # more easily track combinations.\n",
    "        index = 1\n",
    "\n",
    "        # Now iterate over the combinations.\n",
    "        for choose_number in combos.keys():\n",
    "            for selections in combos[choose_number]:\n",
    "\n",
    "                # Create a variable to hold the length.\n",
    "                intermediate_transcript_length = minimum_transcript_length\n",
    "\n",
    "                # Get exactly which introns we're selecting.\n",
    "                for intron_number in selections[0].split(','):\n",
    "                    \n",
    "                    # intron_number adjusted down because of 0-indexing\n",
    "                    # in python.\n",
    "\n",
    "                    # Select the intron and add to the length.\n",
    "                    intermediate_transcript_length += intron_lengths[transcript_id][int(intron_number) - 1]\n",
    "                \n",
    "                # Write to file.\n",
    "                f.write('\\t'.join([\n",
    "                    transcript_id, \n",
    "                    str(minimum_transcript_length), \n",
    "                    str(maximum_transcript_length), \n",
    "                    str(n_exons_in_transcript),\n",
    "                    str(n_introns_in_transcript),\n",
    "                    str(index),\n",
    "                    str(choose_number), \n",
    "                    selections[0], \n",
    "                    str(intermediate_transcript_length)\n",
    "                ]) + '\\n')\n",
    "                \n",
    "                # Increase the index.\n",
    "                index += 1  "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "3d_rna_velocity",
   "language": "python",
   "name": "3d_rna_velocity"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
