{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ec03025f",
   "metadata": {},
   "source": [
    "# GFF Processing Metrics"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0532fda4",
   "metadata": {},
   "source": [
    "Using a processed GFF file (either from NIH or Ensembl) we can apply some metrics to understand the characteristics of our RNA dataset."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1fe57bb",
   "metadata": {},
   "source": [
    "# Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f36d533e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Libraries\n",
    "\n",
    "\n",
    "# Combinations\n",
    "from itertools import combinations\n",
    "\n",
    "# Floor\n",
    "import math\n",
    "\n",
    "# NumPy\n",
    "import numpy as np\n",
    "\n",
    "# Pandas\n",
    "import pandas as pd\n",
    "\n",
    "# Sampling\n",
    "# import random  \n",
    "from random import sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "45cacf5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set the data folder.\n",
    "data_folder = '/media/apollo/Samsung_T5/transfer/mayur/annotations/'\n",
    "# data_folder = '/home/mad1188/rvallsamples/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5b6a005",
   "metadata": {},
   "source": [
    "# NIH File"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fce186d9",
   "metadata": {},
   "source": [
    "First, read in the processed GFF file from NIH."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a1235032",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# NIH file.\n",
    "\n",
    "# Skip the metadata lines (1-9).\n",
    "df = pd.read_csv(\n",
    "    data_folder + 'GCF_000001405.40_GRCh38.p14_genomic.gff.exon.processed',\n",
    "    sep = '\\t'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "bd260327",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>nih_molecule_accession</th>\n",
       "      <th>source</th>\n",
       "      <th>category</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>information</th>\n",
       "      <th>ID</th>\n",
       "      <th>exon_parent</th>\n",
       "      <th>exon_gbkey</th>\n",
       "      <th>exon_gene</th>\n",
       "      <th>exon_transcript_id</th>\n",
       "      <th>predicted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>17369.0</td>\n",
       "      <td>17436.0</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=exon-NR_106918.1-1;Parent=rna-NR_106918.1;D...</td>\n",
       "      <td>exon-NR_106918.1-1</td>\n",
       "      <td>rna-NR_106918.1</td>\n",
       "      <td>precursor_RNA</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>NR_106918.1</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>17369.0</td>\n",
       "      <td>17391.0</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=exon-MIR6859-1-1;Parent=rna-MIR6859-1;Dbxre...</td>\n",
       "      <td>exon-MIR6859-1-1</td>\n",
       "      <td>rna-MIR6859-1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>17409.0</td>\n",
       "      <td>17431.0</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=exon-MIR6859-1-2-1;Parent=rna-MIR6859-1-2;D...</td>\n",
       "      <td>exon-MIR6859-1-2-1</td>\n",
       "      <td>rna-MIR6859-1-2</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>29774.0</td>\n",
       "      <td>30667.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-XR_007065314.1-1;Parent=rna-XR_0070653...</td>\n",
       "      <td>exon-XR_007065314.1-1</td>\n",
       "      <td>rna-XR_007065314.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>XR_007065314.1</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>30976.0</td>\n",
       "      <td>31093.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-XR_007065314.1-2;Parent=rna-XR_0070653...</td>\n",
       "      <td>exon-XR_007065314.1-2</td>\n",
       "      <td>rna-XR_007065314.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>XR_007065314.1</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1698714</th>\n",
       "      <td>NC_012920.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>14149.0</td>\n",
       "      <td>14673.0</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=exon-ND6-1;Parent=rna-ND6;Dbxref=GeneID:454...</td>\n",
       "      <td>exon-ND6-1</td>\n",
       "      <td>rna-ND6</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>ND6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1698715</th>\n",
       "      <td>NC_012920.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>14674.0</td>\n",
       "      <td>14742.0</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=exon-TRNE-1;Parent=rna-TRNE;Dbxref=GeneID:4...</td>\n",
       "      <td>exon-TRNE-1</td>\n",
       "      <td>rna-TRNE</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>TRNE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1698716</th>\n",
       "      <td>NC_012920.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>14747.0</td>\n",
       "      <td>15887.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-CYTB-1;Parent=rna-CYTB;Dbxref=GeneID:4...</td>\n",
       "      <td>exon-CYTB-1</td>\n",
       "      <td>rna-CYTB</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>CYTB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1698717</th>\n",
       "      <td>NC_012920.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>15888.0</td>\n",
       "      <td>15953.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-TRNT-1;Parent=rna-TRNT;Dbxref=GeneID:4...</td>\n",
       "      <td>exon-TRNT-1</td>\n",
       "      <td>rna-TRNT</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>TRNT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1698718</th>\n",
       "      <td>NC_012920.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>15956.0</td>\n",
       "      <td>16023.0</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=exon-TRNP-1;Parent=rna-TRNP;Dbxref=GeneID:4...</td>\n",
       "      <td>exon-TRNP-1</td>\n",
       "      <td>rna-TRNP</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>TRNP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1698719 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        nih_molecule_accession      source category    start     stop strand  \\\n",
       "0                 NC_000001.11  BestRefSeq     exon  17369.0  17436.0      -   \n",
       "1                 NC_000001.11  BestRefSeq     exon  17369.0  17391.0      -   \n",
       "2                 NC_000001.11  BestRefSeq     exon  17409.0  17431.0      -   \n",
       "3                 NC_000001.11      Gnomon     exon  29774.0  30667.0      +   \n",
       "4                 NC_000001.11      Gnomon     exon  30976.0  31093.0      +   \n",
       "...                        ...         ...      ...      ...      ...    ...   \n",
       "1698714            NC_012920.1      RefSeq     exon  14149.0  14673.0      -   \n",
       "1698715            NC_012920.1      RefSeq     exon  14674.0  14742.0      -   \n",
       "1698716            NC_012920.1      RefSeq     exon  14747.0  15887.0      +   \n",
       "1698717            NC_012920.1      RefSeq     exon  15888.0  15953.0      +   \n",
       "1698718            NC_012920.1      RefSeq     exon  15956.0  16023.0      -   \n",
       "\n",
       "                                               information  \\\n",
       "0        ID=exon-NR_106918.1-1;Parent=rna-NR_106918.1;D...   \n",
       "1        ID=exon-MIR6859-1-1;Parent=rna-MIR6859-1;Dbxre...   \n",
       "2        ID=exon-MIR6859-1-2-1;Parent=rna-MIR6859-1-2;D...   \n",
       "3        ID=exon-XR_007065314.1-1;Parent=rna-XR_0070653...   \n",
       "4        ID=exon-XR_007065314.1-2;Parent=rna-XR_0070653...   \n",
       "...                                                    ...   \n",
       "1698714  ID=exon-ND6-1;Parent=rna-ND6;Dbxref=GeneID:454...   \n",
       "1698715  ID=exon-TRNE-1;Parent=rna-TRNE;Dbxref=GeneID:4...   \n",
       "1698716  ID=exon-CYTB-1;Parent=rna-CYTB;Dbxref=GeneID:4...   \n",
       "1698717  ID=exon-TRNT-1;Parent=rna-TRNT;Dbxref=GeneID:4...   \n",
       "1698718  ID=exon-TRNP-1;Parent=rna-TRNP;Dbxref=GeneID:4...   \n",
       "\n",
       "                            ID         exon_parent     exon_gbkey  \\\n",
       "0           exon-NR_106918.1-1     rna-NR_106918.1  precursor_RNA   \n",
       "1             exon-MIR6859-1-1       rna-MIR6859-1          ncRNA   \n",
       "2           exon-MIR6859-1-2-1     rna-MIR6859-1-2          ncRNA   \n",
       "3        exon-XR_007065314.1-1  rna-XR_007065314.1          ncRNA   \n",
       "4        exon-XR_007065314.1-2  rna-XR_007065314.1          ncRNA   \n",
       "...                        ...                 ...            ...   \n",
       "1698714             exon-ND6-1             rna-ND6           mRNA   \n",
       "1698715            exon-TRNE-1            rna-TRNE           tRNA   \n",
       "1698716            exon-CYTB-1            rna-CYTB           mRNA   \n",
       "1698717            exon-TRNT-1            rna-TRNT           tRNA   \n",
       "1698718            exon-TRNP-1            rna-TRNP           tRNA   \n",
       "\n",
       "           exon_gene exon_transcript_id      predicted  \n",
       "0          MIR6859-1        NR_106918.1             no  \n",
       "1          MIR6859-1                NaN  not_available  \n",
       "2          MIR6859-1                NaN  not_available  \n",
       "3        MIR1302-2HG     XR_007065314.1            yes  \n",
       "4        MIR1302-2HG     XR_007065314.1            yes  \n",
       "...              ...                ...            ...  \n",
       "1698714          ND6                NaN  not_available  \n",
       "1698715         TRNE                NaN  not_available  \n",
       "1698716         CYTB                NaN  not_available  \n",
       "1698717         TRNT                NaN  not_available  \n",
       "1698718         TRNP                NaN  not_available  \n",
       "\n",
       "[1698719 rows x 13 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67e2cdd3",
   "metadata": {},
   "source": [
    "# Metrics"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15bd94aa",
   "metadata": {},
   "source": [
    "For the metrics that follow, we will concern ourselves only with exons for which a transcript ID is given (including predicted products)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "6e2a05a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Only want ones with transcripts.\n",
    "has_transcript = df[df['predicted'] != 'not_available'].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c290384c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>nih_molecule_accession</th>\n",
       "      <th>source</th>\n",
       "      <th>category</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>information</th>\n",
       "      <th>ID</th>\n",
       "      <th>exon_parent</th>\n",
       "      <th>exon_gbkey</th>\n",
       "      <th>exon_gene</th>\n",
       "      <th>exon_transcript_id</th>\n",
       "      <th>predicted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>17369.0</td>\n",
       "      <td>17436.0</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=exon-NR_106918.1-1;Parent=rna-NR_106918.1;D...</td>\n",
       "      <td>exon-NR_106918.1-1</td>\n",
       "      <td>rna-NR_106918.1</td>\n",
       "      <td>precursor_RNA</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>NR_106918.1</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>29774.0</td>\n",
       "      <td>30667.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-XR_007065314.1-1;Parent=rna-XR_0070653...</td>\n",
       "      <td>exon-XR_007065314.1-1</td>\n",
       "      <td>rna-XR_007065314.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>XR_007065314.1</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>30976.0</td>\n",
       "      <td>31093.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-XR_007065314.1-2;Parent=rna-XR_0070653...</td>\n",
       "      <td>exon-XR_007065314.1-2</td>\n",
       "      <td>rna-XR_007065314.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>XR_007065314.1</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>34168.0</td>\n",
       "      <td>35418.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-XR_007065314.1-3;Parent=rna-XR_0070653...</td>\n",
       "      <td>exon-XR_007065314.1-3</td>\n",
       "      <td>rna-XR_007065314.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>XR_007065314.1</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>30366.0</td>\n",
       "      <td>30503.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-NR_036051.1-1;Parent=rna-NR_036051.1;D...</td>\n",
       "      <td>exon-NR_036051.1-1</td>\n",
       "      <td>rna-NR_036051.1</td>\n",
       "      <td>precursor_RNA</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>NR_036051.1</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1698677</th>\n",
       "      <td>NC_000024.10</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>57192325.0</td>\n",
       "      <td>57192537.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-XM_017030055.2-1;Parent=rna-XM_0170300...</td>\n",
       "      <td>exon-XM_017030055.2-1</td>\n",
       "      <td>rna-XM_017030055.2</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>IL9R</td>\n",
       "      <td>XM_017030055.2</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1698678</th>\n",
       "      <td>NC_000024.10</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>57194043.0</td>\n",
       "      <td>57194127.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-XM_017030055.2-2;Parent=rna-XM_0170300...</td>\n",
       "      <td>exon-XM_017030055.2-2</td>\n",
       "      <td>rna-XM_017030055.2</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>IL9R</td>\n",
       "      <td>XM_017030055.2</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1698679</th>\n",
       "      <td>NC_000024.10</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>57196336.0</td>\n",
       "      <td>57197337.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=exon-XM_017030055.2-3;Parent=rna-XM_0170300...</td>\n",
       "      <td>exon-XM_017030055.2-3</td>\n",
       "      <td>rna-XM_017030055.2</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>IL9R</td>\n",
       "      <td>XM_017030055.2</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1698680</th>\n",
       "      <td>NC_000024.10</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>57203182.0</td>\n",
       "      <td>57203350.0</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=exon-NR_138048.1-2-1;Parent=rna-NR_138048.1...</td>\n",
       "      <td>exon-NR_138048.1-2-1</td>\n",
       "      <td>rna-NR_138048.1-2</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>WASIR1</td>\n",
       "      <td>NR_138048.1</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1698681</th>\n",
       "      <td>NC_000024.10</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>57201084.0</td>\n",
       "      <td>57202145.0</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=exon-NR_138048.1-2-2;Parent=rna-NR_138048.1...</td>\n",
       "      <td>exon-NR_138048.1-2-2</td>\n",
       "      <td>rna-NR_138048.1-2</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>WASIR1</td>\n",
       "      <td>NR_138048.1</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1694667 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        nih_molecule_accession      source category       start        stop  \\\n",
       "0                 NC_000001.11  BestRefSeq     exon     17369.0     17436.0   \n",
       "3                 NC_000001.11      Gnomon     exon     29774.0     30667.0   \n",
       "4                 NC_000001.11      Gnomon     exon     30976.0     31093.0   \n",
       "5                 NC_000001.11      Gnomon     exon     34168.0     35418.0   \n",
       "6                 NC_000001.11  BestRefSeq     exon     30366.0     30503.0   \n",
       "...                        ...         ...      ...         ...         ...   \n",
       "1698677           NC_000024.10      Gnomon     exon  57192325.0  57192537.0   \n",
       "1698678           NC_000024.10      Gnomon     exon  57194043.0  57194127.0   \n",
       "1698679           NC_000024.10      Gnomon     exon  57196336.0  57197337.0   \n",
       "1698680           NC_000024.10  BestRefSeq     exon  57203182.0  57203350.0   \n",
       "1698681           NC_000024.10  BestRefSeq     exon  57201084.0  57202145.0   \n",
       "\n",
       "        strand                                        information  \\\n",
       "0            -  ID=exon-NR_106918.1-1;Parent=rna-NR_106918.1;D...   \n",
       "3            +  ID=exon-XR_007065314.1-1;Parent=rna-XR_0070653...   \n",
       "4            +  ID=exon-XR_007065314.1-2;Parent=rna-XR_0070653...   \n",
       "5            +  ID=exon-XR_007065314.1-3;Parent=rna-XR_0070653...   \n",
       "6            +  ID=exon-NR_036051.1-1;Parent=rna-NR_036051.1;D...   \n",
       "...        ...                                                ...   \n",
       "1698677      +  ID=exon-XM_017030055.2-1;Parent=rna-XM_0170300...   \n",
       "1698678      +  ID=exon-XM_017030055.2-2;Parent=rna-XM_0170300...   \n",
       "1698679      +  ID=exon-XM_017030055.2-3;Parent=rna-XM_0170300...   \n",
       "1698680      -  ID=exon-NR_138048.1-2-1;Parent=rna-NR_138048.1...   \n",
       "1698681      -  ID=exon-NR_138048.1-2-2;Parent=rna-NR_138048.1...   \n",
       "\n",
       "                            ID         exon_parent     exon_gbkey  \\\n",
       "0           exon-NR_106918.1-1     rna-NR_106918.1  precursor_RNA   \n",
       "3        exon-XR_007065314.1-1  rna-XR_007065314.1          ncRNA   \n",
       "4        exon-XR_007065314.1-2  rna-XR_007065314.1          ncRNA   \n",
       "5        exon-XR_007065314.1-3  rna-XR_007065314.1          ncRNA   \n",
       "6           exon-NR_036051.1-1     rna-NR_036051.1  precursor_RNA   \n",
       "...                        ...                 ...            ...   \n",
       "1698677  exon-XM_017030055.2-1  rna-XM_017030055.2           mRNA   \n",
       "1698678  exon-XM_017030055.2-2  rna-XM_017030055.2           mRNA   \n",
       "1698679  exon-XM_017030055.2-3  rna-XM_017030055.2           mRNA   \n",
       "1698680   exon-NR_138048.1-2-1   rna-NR_138048.1-2          ncRNA   \n",
       "1698681   exon-NR_138048.1-2-2   rna-NR_138048.1-2          ncRNA   \n",
       "\n",
       "           exon_gene exon_transcript_id predicted  \n",
       "0          MIR6859-1        NR_106918.1        no  \n",
       "3        MIR1302-2HG     XR_007065314.1       yes  \n",
       "4        MIR1302-2HG     XR_007065314.1       yes  \n",
       "5        MIR1302-2HG     XR_007065314.1       yes  \n",
       "6          MIR1302-2        NR_036051.1        no  \n",
       "...              ...                ...       ...  \n",
       "1698677         IL9R     XM_017030055.2       yes  \n",
       "1698678         IL9R     XM_017030055.2       yes  \n",
       "1698679         IL9R     XM_017030055.2       yes  \n",
       "1698680       WASIR1        NR_138048.1        no  \n",
       "1698681       WASIR1        NR_138048.1        no  \n",
       "\n",
       "[1694667 rows x 13 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "has_transcript"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6582c78f",
   "metadata": {},
   "source": [
    "First, we will see the breakdown of the various types of RNA.  Note that we are not able to discriminate miRNA, tRNA, etc... from each other with this processing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "b9c33ee1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Group and count.\n",
    "\n",
    "# Keep DataFrame, but we can't use\n",
    "# .to_frame() because of a header naming\n",
    "# error, see https://github.com/pandas-dev/pandas/issues/6618\n",
    "\n",
    "# Source: https://stackoverflow.com/a/42324086\n",
    "category_counts = has_transcript.groupby('exon_gbkey', as_index = False).size()\n",
    "category_counts = category_counts.rename(\n",
    "    columns = {\n",
    "        \"size\": \"n_category\"\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "81c478b9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon_gbkey</th>\n",
       "      <th>n_category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>mRNA</td>\n",
       "      <td>1504043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>misc_RNA</td>\n",
       "      <td>114898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ncRNA</td>\n",
       "      <td>73782</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>precursor_RNA</td>\n",
       "      <td>1915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rRNA</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      exon_gbkey  n_category\n",
       "0           mRNA     1504043\n",
       "1       misc_RNA      114898\n",
       "2          ncRNA       73782\n",
       "3  precursor_RNA        1915\n",
       "4           rRNA          29"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "category_counts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6345246",
   "metadata": {},
   "source": [
    "What about genes?  Create a small function to work on the groups."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "0638fb82",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Helper function\n",
    "def category_helper(df_temp):\n",
    "    \n",
    "    # Simply count the number of genes.\n",
    "    return(len(set(df_temp.exon_gene)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e32e8537",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_19944/458226104.py:2: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  grouped_and_counted = has_transcript.groupby(['exon_gbkey']).apply(lambda x: category_helper(x))\n"
     ]
    }
   ],
   "source": [
    "# Group and count unique genes per category.\n",
    "grouped_and_counted = has_transcript.groupby(['exon_gbkey']).apply(lambda x: category_helper(x))\n",
    "\n",
    "# We want a DataFrame.\n",
    "grouped_and_counted = pd.DataFrame(\n",
    "    {\n",
    "        'exon_gbkey': list(grouped_and_counted.keys()),\n",
    "        'n_unique_genes': list(grouped_and_counted.values)\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "4a5d2dd4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon_gbkey</th>\n",
       "      <th>n_unique_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>mRNA</td>\n",
       "      <td>19617</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>misc_RNA</td>\n",
       "      <td>3831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ncRNA</td>\n",
       "      <td>12403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>precursor_RNA</td>\n",
       "      <td>1913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rRNA</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      exon_gbkey  n_unique_genes\n",
       "0           mRNA           19617\n",
       "1       misc_RNA            3831\n",
       "2          ncRNA           12403\n",
       "3  precursor_RNA            1913\n",
       "4           rRNA              29"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grouped_and_counted"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb62cdb0",
   "metadata": {},
   "source": [
    "These numbers look about right, so let's move on to predicted vs. non-predicted."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "71d17d5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the two letter codes and make a nice dataframe.\n",
    "\n",
    "# Source: https://www.ncbi.nlm.nih.gov/books/NBK21091/table/ch18.T.refseq_accession_numbers_and_mole/?report=objectonly\n",
    "two_letter_codes_df = pd.DataFrame(\n",
    "    data = {\n",
    "        'category': [i.split('_')[0] for i in list(has_transcript.exon_transcript_id)]\n",
    "    }\n",
    ")\n",
    "\n",
    "two_letter_codes_df = two_letter_codes_df.groupby('category', as_index = False).size()\n",
    "two_letter_codes_df = two_letter_codes_df.rename(\n",
    "    columns = {\n",
    "        \"size\": \"n_category\"\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "273b7c1b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>n_category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NM</td>\n",
       "      <td>817181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NR</td>\n",
       "      <td>135370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>XM</td>\n",
       "      <td>686862</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XR</td>\n",
       "      <td>55254</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  category  n_category\n",
       "0       NM      817181\n",
       "1       NR      135370\n",
       "2       XM      686862\n",
       "3       XR       55254"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "two_letter_codes_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c05085b",
   "metadata": {},
   "source": [
    "Let's get a feel for exon counts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "78ac5836",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Keep DataFrame.\n",
    "# Source: https://stackoverflow.com/a/42324086\n",
    "exon_counts = has_transcript.groupby(['exon_gene', 'exon_transcript_id'], as_index = False).size()\n",
    "exon_counts = exon_counts.rename(\n",
    "    columns = {\n",
    "        \"size\": \"n_exons\"\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "f885db93",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon_gene</th>\n",
       "      <th>exon_transcript_id</th>\n",
       "      <th>n_exons</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>NM_130786.4</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>NR_015380.2</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>NM_001198818.2</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>NM_001198819.2</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>NM_001198820.2</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139214</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>XM_047417329.1</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139215</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>XM_047417333.1</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139216</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>XM_047417337.1</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139217</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>XM_047417341.1</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139218</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>XM_047417342.1</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>139219 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       exon_gene exon_transcript_id  n_exons\n",
       "0           A1BG        NM_130786.4        8\n",
       "1       A1BG-AS1        NR_015380.2        4\n",
       "2           A1CF     NM_001198818.2       14\n",
       "3           A1CF     NM_001198819.2       15\n",
       "4           A1CF     NM_001198820.2       14\n",
       "...          ...                ...      ...\n",
       "139214      ZZZ3     XM_047417329.1       14\n",
       "139215      ZZZ3     XM_047417333.1       13\n",
       "139216      ZZZ3     XM_047417337.1       14\n",
       "139217      ZZZ3     XM_047417341.1       13\n",
       "139218      ZZZ3     XM_047417342.1       13\n",
       "\n",
       "[139219 rows x 3 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon_counts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8051eb54",
   "metadata": {},
   "source": [
    "We can immediately identify the number of introns as well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "1ebfd24b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add the intron counts.\n",
    "exon_counts['n_introns'] = exon_counts['n_exons']-1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "dbe198de",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon_gene</th>\n",
       "      <th>exon_transcript_id</th>\n",
       "      <th>n_exons</th>\n",
       "      <th>n_introns</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>NM_130786.4</td>\n",
       "      <td>8</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>NR_015380.2</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>NM_001198818.2</td>\n",
       "      <td>14</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>NM_001198819.2</td>\n",
       "      <td>15</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>NM_001198820.2</td>\n",
       "      <td>14</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139214</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>XM_047417329.1</td>\n",
       "      <td>14</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139215</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>XM_047417333.1</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139216</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>XM_047417337.1</td>\n",
       "      <td>14</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139217</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>XM_047417341.1</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139218</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>XM_047417342.1</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>139219 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       exon_gene exon_transcript_id  n_exons  n_introns\n",
       "0           A1BG        NM_130786.4        8          7\n",
       "1       A1BG-AS1        NR_015380.2        4          3\n",
       "2           A1CF     NM_001198818.2       14         13\n",
       "3           A1CF     NM_001198819.2       15         14\n",
       "4           A1CF     NM_001198820.2       14         13\n",
       "...          ...                ...      ...        ...\n",
       "139214      ZZZ3     XM_047417329.1       14         13\n",
       "139215      ZZZ3     XM_047417333.1       13         12\n",
       "139216      ZZZ3     XM_047417337.1       14         13\n",
       "139217      ZZZ3     XM_047417341.1       13         12\n",
       "139218      ZZZ3     XM_047417342.1       13         12\n",
       "\n",
       "[139219 rows x 4 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "3c8f3fe5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # To file.\n",
    "# exon_counts.to_csv(\n",
    "#     'exon_intron_counts.tsv',\n",
    "#     index = False,\n",
    "#     sep = '\\t',\n",
    "# )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "baaaf17f",
   "metadata": {},
   "source": [
    "Make a nice table to show the distribution (and therefore necessary work to find transcript lengths) of introns."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "374d7fe7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Group to see the distribution of \n",
    "# intron counts.\n",
    "intron_counts = exon_counts.groupby(['n_introns'], as_index = False).size()\n",
    "intron_counts = intron_counts.rename(\n",
    "    columns = {\n",
    "        \"size\": \"n_in_group\"\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "a4b35f85",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_introns</th>\n",
       "      <th>n_in_group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>5067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>5352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>11871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>10234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>9300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>312</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>316</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>335</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>358</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>362</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>152 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     n_introns  n_in_group\n",
       "0            0        5067\n",
       "1            1        5352\n",
       "2            2       11871\n",
       "3            3       10234\n",
       "4            4        9300\n",
       "..         ...         ...\n",
       "147        312           2\n",
       "148        316           1\n",
       "149        335           2\n",
       "150        358           1\n",
       "151        362           1\n",
       "\n",
       "[152 rows x 2 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intron_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "be7813e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # To file.\n",
    "# intron_counts.to_csv(\n",
    "#     'intron_counts.tsv',\n",
    "#     index = False,\n",
    "#     sep = '\\t'\n",
    "# )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33aa2710",
   "metadata": {},
   "source": [
    "Pick a gene to check the quality of the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "90d32cf1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon_gene</th>\n",
       "      <th>exon_transcript_id</th>\n",
       "      <th>n_exons</th>\n",
       "      <th>n_introns</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>120937</th>\n",
       "      <td>TET2</td>\n",
       "      <td>NM_001127208.3</td>\n",
       "      <td>11</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120938</th>\n",
       "      <td>TET2</td>\n",
       "      <td>NM_017628.4</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120939</th>\n",
       "      <td>TET2</td>\n",
       "      <td>XM_005263082.4</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120940</th>\n",
       "      <td>TET2</td>\n",
       "      <td>XM_006714242.4</td>\n",
       "      <td>9</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120941</th>\n",
       "      <td>TET2</td>\n",
       "      <td>XM_024454102.2</td>\n",
       "      <td>12</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120942</th>\n",
       "      <td>TET2</td>\n",
       "      <td>XM_024454103.2</td>\n",
       "      <td>11</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120943</th>\n",
       "      <td>TET2</td>\n",
       "      <td>XM_047415839.1</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120944</th>\n",
       "      <td>TET2</td>\n",
       "      <td>XM_047415840.1</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120945</th>\n",
       "      <td>TET2</td>\n",
       "      <td>XM_047415841.1</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120946</th>\n",
       "      <td>TET2</td>\n",
       "      <td>XM_047415842.1</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120947</th>\n",
       "      <td>TET2</td>\n",
       "      <td>XR_007057933.1</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       exon_gene exon_transcript_id  n_exons  n_introns\n",
       "120937      TET2     NM_001127208.3       11         10\n",
       "120938      TET2        NM_017628.4        3          2\n",
       "120939      TET2     XM_005263082.4       10          9\n",
       "120940      TET2     XM_006714242.4        9          8\n",
       "120941      TET2     XM_024454102.2       12         11\n",
       "120942      TET2     XM_024454103.2       11         10\n",
       "120943      TET2     XM_047415839.1        6          5\n",
       "120944      TET2     XM_047415840.1        6          5\n",
       "120945      TET2     XM_047415841.1        5          4\n",
       "120946      TET2     XM_047415842.1        5          4\n",
       "120947      TET2     XR_007057933.1        5          4"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TET2 check.\n",
    "exon_counts[exon_counts['exon_gene'] == 'TET2']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd35d8af",
   "metadata": {},
   "source": [
    "Optionally, sort by the number of introns and write out."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "41460ae8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon_gene</th>\n",
       "      <th>exon_transcript_id</th>\n",
       "      <th>n_exons</th>\n",
       "      <th>n_introns</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>NM_130786.4</td>\n",
       "      <td>8</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>NR_015380.2</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>NM_001370130.1</td>\n",
       "      <td>12</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>NM_001370131.1</td>\n",
       "      <td>12</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>XM_011539729.4</td>\n",
       "      <td>12</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139205</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>NM_001376154.1</td>\n",
       "      <td>15</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139208</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>NM_015534.6</td>\n",
       "      <td>15</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139209</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>NR_164775.1</td>\n",
       "      <td>15</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139210</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>NR_164776.1</td>\n",
       "      <td>15</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139211</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>NR_164777.1</td>\n",
       "      <td>15</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>139219 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       exon_gene exon_transcript_id  n_exons  n_introns\n",
       "0           A1BG        NM_130786.4        8          7\n",
       "1       A1BG-AS1        NR_015380.2        4          3\n",
       "5           A1CF     NM_001370130.1       12         11\n",
       "6           A1CF     NM_001370131.1       12         11\n",
       "11          A1CF     XM_011539729.4       12         11\n",
       "...          ...                ...      ...        ...\n",
       "139205      ZZZ3     NM_001376154.1       15         14\n",
       "139208      ZZZ3        NM_015534.6       15         14\n",
       "139209      ZZZ3        NR_164775.1       15         14\n",
       "139210      ZZZ3        NR_164776.1       15         14\n",
       "139211      ZZZ3        NR_164777.1       15         14\n",
       "\n",
       "[139219 rows x 4 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Sort on number of introns.\n",
    "exon_counts.sort_values(by = ['exon_gene', 'n_introns'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "519f9f29",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write out..."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fbb6ae7e",
   "metadata": {},
   "source": [
    "How about truly binary splice/unspliced genes (useful in scVelo)?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "69540626",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon_gene</th>\n",
       "      <th>exon_transcript_id</th>\n",
       "      <th>n_exons</th>\n",
       "      <th>n_introns</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>NR_137424.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>NR_137425.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>A2ML1-AS1</td>\n",
       "      <td>NR_046715.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>NR_146459.3</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>AADACL4</td>\n",
       "      <td>XM_017001153.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139093</th>\n",
       "      <td>ZSWIM1</td>\n",
       "      <td>XM_005260611.5</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139096</th>\n",
       "      <td>ZSWIM3</td>\n",
       "      <td>NM_080752.4</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139146</th>\n",
       "      <td>ZSWIM9</td>\n",
       "      <td>XM_047438788.1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139167</th>\n",
       "      <td>ZXDA</td>\n",
       "      <td>NM_007156.5</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139168</th>\n",
       "      <td>ZXDB</td>\n",
       "      <td>NM_007157.4</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10419 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        exon_gene exon_transcript_id  n_exons  n_introns\n",
       "21        A2M-AS1        NR_137424.1        2          1\n",
       "22        A2M-AS1        NR_137425.1        2          1\n",
       "29      A2ML1-AS1        NR_046715.1        2          1\n",
       "33         A4GALT        NR_146459.3        2          1\n",
       "65        AADACL4     XM_017001153.1        2          1\n",
       "...           ...                ...      ...        ...\n",
       "139093     ZSWIM1     XM_005260611.5        2          1\n",
       "139096     ZSWIM3        NM_080752.4        2          1\n",
       "139146     ZSWIM9     XM_047438788.1        2          1\n",
       "139167       ZXDA        NM_007156.5        1          0\n",
       "139168       ZXDB        NM_007157.4        1          0\n",
       "\n",
       "[10419 rows x 4 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# True binaries.\n",
    "exon_counts[exon_counts.n_introns <= 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e04401c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write out..."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "3d_rna_velocity",
   "language": "python",
   "name": "3d_rna_velocity"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
